core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java

```diff
@@ -665,36 +644,54 @@ public InputStream openExistingFile(String path, OpenOptions options) throws IOE
 
   @Override
   public boolean renameDirectory(String src, String dst) throws IOException {
+    if (exists(dst)) {
+      LOG.error("Unable to rename {} to {} because destination already exists.", src, dst);
+      return false;
+    }
+    // Use a global delete buffer, in order to merge delete object requests
+    DeleteBuffer deleteBuffer = new DeleteBuffer();
+    boolean result = renameDirectoryInternal(src, dst, deleteBuffer);
+    int fileDeleted = deleteBuffer.getResult().size();
+    if (fileDeleted != deleteBuffer.mEntriesAdded) {
+      LOG.warn("Failed to rename directory, successfully deleted {} files out of {}.",
+          fileDeleted, deleteBuffer.mEntriesAdded);
+      return false;
+    }
+    return result;
+  }
+
+  private boolean renameDirectoryInternal(String src, String dst, DeleteBuffer deleteBuffer)
+      throws IOException {
     UfsStatus[] children = listInternal(src, ListOptions.defaults());
     if (children == null) {
       LOG.error("Failed to list directory {}, aborting rename.", src);
       return false;
     }
-    if (exists(dst)) {
-      LOG.error("Unable to rename {} to {} because destination already exists.", src, dst);
-      return false;
-    }
     // Source exists and is a directory, and destination does not exist
     // Rename the source folder first
-    if (!copyObject(stripPrefixIfPresent(convertToFolderName(src)),
-        stripPrefixIfPresent(convertToFolderName(dst)))) {
+    String srcKey = stripPrefixIfPresent(convertToFolderName(src));
+    if (!copyObject(srcKey, stripPrefixIfPresent(convertToFolderName(dst)))) {
       return false;
     }
+    deleteBuffer.add(srcKey);
+
     // Rename each child in the src folder to destination/child
     // a. Since renames are a copy operation, files are added to a buffer and processed concurrently
     // b. Pseudo-directories are metadata only operations are not added to the buffer
-    RenameBuffer buffer = new RenameBuffer();
+    RenameBuffer buffer = new RenameBuffer(deleteBuffer);
     for (UfsStatus child : children) {
       String childSrcPath = PathUtils.concatPath(src, child.getName());
       String childDstPath = PathUtils.concatPath(dst, child.getName());
       if (child.isDirectory()) {
         // Recursive call
-        if (!renameDirectory(childSrcPath, childDstPath)) {
+        if (!renameDirectoryInternal(childSrcPath, childDstPath, deleteBuffer)) {
           LOG.error("Failed to rename path {} to {}, aborting rename.", childSrcPath, childDstPath);
           return false;
         }
       } else {
         buffer.add(new Pair<>(childSrcPath, childDstPath));
+        // Delete src
+        deleteBuffer.add(stripPrefixIfPresent(childSrcPath));
```

Comment:
same here

Reply:
this still stands, as buffer.add -> rename op could been yet to trigger, but deleteBuffer.add -> triggers the deletion.

Reply:
fix.